src/ddmd/func.d

```diff
@@ -3451,13 +3451,20 @@ extern (C++) class FuncDeclaration : Declaration
             {
                 Dsymbol st = tret.toDsymbol(null);
                 //printf("\t\treturning class/struct %s\n", tret.toChars());
-                for (Dsymbol s = st.parent; s; s = s.parent)
+
+                auto ad = st.isAggregateDeclaration();
+                assert(ad !is null);
+                // A closure is only needed when returning a _non-static_ struct/class.
+                if ((ad.storage_class & STCstatic) == 0)
                 {
-                    //printf("\t\t\tparent = %s %s\n", s.kind(), s.toChars());
-                    if (s == this)
+                    for (Dsymbol s = st.parent; s; s = s.parent)
                     {
-                        //printf("\t\treturning local %s\n", st.toChars());
-                        goto Lyes;
+                        //printf("\t\t\tparent = %s %s\n", s.kind(), s.toChars());
+                        if (s == this)
+                        {
+                            //printf("\t\treturning local %s\n", st.toChars());
+                            goto Lyes;
```

Comment:
I'm going to try removing that whole case and test it on Weka's codebase. It could be a good optimization to get rid of the closure alloc, and would allow more `@nogc`. On the other hand, it is perhaps better when people are pointed (by `@nogc` compile error) towards adding `static` to a struct that can be static...